internal/internal_workflow.go

```diff
@@ -1347,3 +1356,59 @@ func (h *queryHandler) execute(input []byte) (result []byte, err error) {
 	}
 	return result, err
 }
+
+// Add adds delta, which may be negative, to the WaitGroup counter.
+// If the counter becomes zero, all goroutines blocked on Wait are released.
+// If the counter goes negative, Add panics.
+//
+// Note that calls with a positive delta that occur when the counter is zero
+// must happen before a Wait. Calls with a negative delta, or calls with a
+// positive delta that start when the counter is greater than zero, may happen
+// at any time.
+// Typically this means the calls to Add should execute before the statement
+// creating the goroutine or other event to be waited for.
+// If a WaitGroup is reused to wait for several independent sets of events,
+// new Add calls must happen after all previous Wait calls have returned.
+//
+// param delta int32 -> the value to increment the WaitGroup counter by
+func (wg *waitGroupImpl) Add(delta int32) {
+	if (wg.waiting && delta > 0) && (wg.n == 0) {
+		panic("WaitGroup misuse: Add called concurrently with Wait")
+	}
+
+	wg.n = wg.n + delta
+	if wg.n < 0 {
+		panic("negative WaitGroup counter")
+	}
+	if (wg.n > 0) || (!wg.waiting) {
+		return
+	}
+	if wg.n == 0 {
+		wg.settable.Set(false, nil)
+	}
+}
+
+// Done decrements the WaitGroup counter by 1, indicating
+// that a coroutine in the WaitGroup has completed
+func (wg *waitGroupImpl) Done() {
+	wg.Add(-1)
+}
+
+// Wait blocks and waits for specified number of couritines to
+// finish executing and then unblocks once the counter has reached 0.
+//
+// param ctx Context -> workflow context
+func (wg *waitGroupImpl) Wait(ctx Context) {
+	if wg.n <= 0 {
+		return
+	}
+	if wg.waiting {
+		panic("WaitGroup is reused before previous Wait has returned")
+	}
+
+	wg.waiting = true
+	if err := wg.future.Get(ctx, &wg.waiting); err != nil {
+		panic(err)
+	}
+	wg.future, wg.settable = NewFuture(ctx)
```

Comment:
I would lazily initialize it in the Add to avoid unnecessary creation in the cases when the WaitGroup is not reused.

Reply:
I like the change.  This would make it such that I would have to either store the context or pass a context to the Done() and Add() methods.  Would this make it less straightforward to use?

Reply:
Agree. Adding context to Add is excessive.

Reply:
So I guess the alternative would be to store the context used to initialize the Wait Group, but this would keep one from adding custom context properties such as timeout when waiting on the future to unblock.